### DISCUSSION TITLE
How did you make the canvas so effortlessly manipulatable?

### DISCUSSION THREAD

**Discussion by edthoms** [bitbof/klecks#49] [Q&A] [answered]
I have been researching the internet for weeks on how to make a canvas that is as manipulatable as [klecks's](https://www.dropbox.com/s/27hbcgmychi5kwj/canvasManipulation.mov?dl=0) is. But have sadly come up empty handed. How did you pull it off? lol

Do you keep an array of paths for every path that gets drawn to the canvas and then adjust its positioning to move accordingly with the canvas as it gets rotated/moved? (Using matrix transforms, maybe?) 

Or do you commit every drawn path to the canvas, as if it was paper, thus not needing to keep track of each path? If so, can I ask how you get the canvas to behave as if it was paper? 

Something else that I have experienced is that when I rotate the canvas, input is read wrong when it is rotated and I click/draw on it. As if the "input reader" of the canvas remains at the same position and size, but the visible "paper-looking" canvas is rotated. How do you avoid that from happening, do you reset the transformation matrix of the canvas to the new rotation angle and scale multiplier??

Any light you can shed would be greatly appreciated. Thanks.

**Comment by bitbof** [comment, marked as answer]
Everything is committed to the canvas (each layer its own canvas), this makes drawing a transformed canvas easy. Klecks keeps an array of past actions, but only for undo/redo. Some brushes store the original input events (old approach, less memory but slow), while others simply store the resulting pixel data via ImageData (new approach, faster for something like the smudge brush).

KlCanvasWorkspace (the component that renders the canvas) handles all of the transformation/rotation. The brushes themselves don't really know about rotation or zoom of the canvas. They assume the canvas coordinate system.

KlCanvasWorkspace works with two different coordinate systems:
- Workspace coordinate System - basically the viewport that the user sees (within which the canvas is displayed rotated, transformed), and the coordinate system that input comes in from PointerEvents.
- Canvas coordinate system - coordinate system of the canvas (no rotation, same as a CanvasRenderingContext2D initially)

Transformation between booth systems done with: `workspaceToCanvasCoord(...)` and `canvasToWorkspaceCoord(...)` in KlCanvasWorkspace.

PointerEvents come in the workspace coordinate system, the workspace transforms (matrix transform) them to canvas coordinate system before handing them to the brushes and other tools. Some tools like Shape Tool also get access to the rotation of the workspace, because this affects how shapes are drawn.

> the "input reader" of the canvas remains at the same position and size, but the visible "paper-looking" canvas is rotated.

Your input is in a different coordinate system than the canvas itself. You need to transform input to the coord system of your canvas before drawing.

Tip: Being 100% specific about your coordinate systems will make you avoid confusion and bugs. I.e. where is the origin, what is the orientation, what is the size of the unit vector (in relation to a pixel, etc).

**Reply by edthoms** [reply]
Quick question, when the paper is moved in klecks, is that paper a `<canvas>` element being moved, panned, and rotated inside a `<div>`? or is that a `rect` inside a `<canvas>` element?

If it's the former, a `<canvas>` inside a `<div>`, how do you suggest I go about transforming the input coordinates to the `<canvas>` coordinate system? Using `setTransform() /getTransform()`? 

Thanks, I'm a newbie at this : )

**Reply by bitbof** [reply]
It's canvases (one per layer) that are drawn on a canvas. Check out KlCanvasWorkspace, or inspect the page with dev tools.
Years ago Kleki transformed divs and canvas objects with CSS. It got too glitchy and messy over time. If you're starting fresh it might be even better to use WebGL, such as https://github.com/marcello3d/lascaux-sketch

**Reply by edthoms** [reply]
@bitbof Excellent, mate. Thanks so much for your help.

**Comment by edthoms** [comment]
@bitbof Thank you so much.
